### PR TITLE
Add ctrl/shift modifiers to footpath tool

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Various tools in the game allow the use of ctrl/shift modifier keys to adjust the base height of element placement. This PR brings this functionality to the footpath tool.

~~Currently, this is still mostly a prototype. It works, but the code could probably do with a cleanup before this is merged. It also currently allows placing paths at half heights, which it shouldn't.~~

https://github.com/user-attachments/assets/8bede132-315f-4912-baca-41ec57438b73

### CODE REVIEW

**Comment by fidwell** [comment]
Very cool! One thing I noticed is that in your video, you're placing paths at invalid heights. Is the cheat on? If not, that'll need to be addressed, assuming it isn't already on your list.

**Comment by AaronVanGeffen** [comment]
> Very cool! One thing I noticed is that in your video, you're placing paths at invalid heights. Is the cheat on? If not, that'll need to be addressed, assuming it isn't already on your list.

You're right, it currently allows placing at invalid heights, owing to basing the height code on that of the ride and scenery windows, which do allow building at those heights.

I'd like to clean up the code and integrate the existing code paths more into the one introduced in this PR, such that only full heights are allowed again.

Another problem with the current implementation is that the 'modifier' baseZ is used even if the modifier key isn't being held, haha.

File: src/openrct2-ui/windows/Footpath.cpp [hunk 1]
```diff
@@ -891,44 +1065,33 @@ static constexpr uint8_t ConstructionPreviewImages[][4] = {
 
             FootpathProvisionalUpdate();
 
-            const auto info = GetMapCoordinatesFromPos(
-                screenCoords, EnumsToFlags(ViewportInteractionItem::Terrain, ViewportInteractionItem::Footpath));
-
-            if (info.SpriteType == ViewportInteractionItem::None)
-            {
+            // TODO: integrate code path below
```
Comment (AaronVanGeffen) [inline src/openrct2-ui/windows/Footpath.cpp:1068]: Outdated comment; can be removed.

**Review by karst — CHANGES_REQUESTED** [review]
It feels great! 
Issues I have found:
- The Ctrl option doesn't snap to scenery. Scenery does snap to the height of paths and this is an inconsistency.
- There's a bug happening when you have a path in the foreground on a higher level, it will bring the path in the background to that level, on the background tile, without holding shift or ctrl. 
image
- The level grid you get when holding shift is offset by a unit, when the "construct bridge or tunnel footpath" is set to upwards.
- Tooltip for this mode is called "Construct footpath on land" should probably be changed to something like "Draw footpath". And the other one should be renamed too for consistency sake.

**Comment by AaronVanGeffen** [comment]
Thanks for testing, @karst!

> The Ctrl option doesn't snap to scenery. Scenery does snap to the height of paths and this is an inconsistency.

This should now be fixed. I was using the interaction flags used by the ride construction window, but I have switched to the ones from the scenery window.

> There's a bug happening when you have a path in the foreground on a higher level, it will bring the path in the background to that level, on the background tile, without holding shift or ctrl.

Not sure why this is happening yet. Will investigate.

> The level grid you get when holding shift is offset by a unit, when the "construct bridge or tunnel footpath" is set to upwards.

I'm not sure why that is the case, but the virtual floor code is also not touched by this PR. If you believe this is a bug, could you file an issue for it?

> Tooltip for this mode is called "Construct footpath on land" should probably be changed to something like "Draw footpath". And the other one should be renamed too for consistency sake.

Personally, I'd like to leave the tooltips as-is. The new behaviour is additive for advanced players. Both modes still work as they used to without the modifier keys. I'd rather avoid confusing new players.

**Comment by karst** [comment]
> I'm not sure why that is the case, but the virtual floor code is also not touched by this PR. If you believe this is a bug, could you file an issue for it?

Ah yes, this seems to be present in develop, too.

**Comment by AaronVanGeffen** [comment]
> There's a bug happening when you have a path in the foreground on a higher level, it will bring the path in the background to that level, on the background tile, without holding shift or ctrl.

This should be fixed as of the latest commit.

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 2]
```diff
@@ -2928,6 +2928,7 @@ static Widget _rideConstructionWidgets[] = {
         {
             if (im.IsModifierKeyPressed(ModifierKey::ctrl))
             {
+                // TODO: use ViewportInteractionItem flags?
```
Comment (AaronVanGeffen) [inline src/openrct2-ui/windows/RideConstruction.cpp:2931]: #22607

**Review by karst — APPROVED** [review]
All previously mentioned issues (except for the bug in develop) have been tackled. All seems to work correctly. Nice work!

**Review by duncanspumpkin — APPROVED** [review]
Code looks sensible. Trusting @karst  on the playability.

**Pull request merged**